tests/form-filling.js

```diff
@@ -0,0 +1,101 @@
+    name = await element.getAttribute('name');
+    cssAttribute = 'name';
+  }
+  return [cssAttribute, name];
+};
+
+const getVisibleAnswers = async () => {
+  const answers = await element.all(by.css('.current input:not([disabled],.ignore)'));
+  const visibleAnswers = await answers.map(async (answer) => {
+    const [cssAttribute, name] = await getAttributeAndName(answer);
+    const type = await answer.getAttribute('type');
+    return { 
+      name: name,
+      type: type,
+      css: `input[${cssAttribute}="${name}"]`
+    };
+  });
+  return await Promise.all(visibleAnswers);
+};
+
+const checkbox = (answer,answerVal) => {
+  const checks = answerVal.split();
+  checks.forEach(async (check) => {
+    await helper.clickElementNative(element(by.css(`${answer.css}[value=${check}]`)));
+  });
+};
+
+const radio = async (answer, answerVal) => {
+  await helper.clickElementNative(element(by.css(`${answer.css}[value="${answerVal}"]`)));
+};
+
+const date = async (answer, answerVal) => {
+  const css = `${answer.css} + div input`;
+  await element(by.css(css)).sendKeys(answerVal);
+};
+
+const defaultAction = async (answer, answerVal) => {
+  await element(by.css(answer.css)).sendKeys(answerVal);
+};
+
+const answerActions = {checkbox ,radio, date};
+
+const repeats = (pathKeys) => {
+  const newKeys = pathKeys;
+  const removedIndexes = {};
+  Object.keys(newKeys).forEach((key)=>{
+    const repeatNum = key.match(/(\.\d\.)/g);
+    if(repeatNum){
+      const stripped = key.replace(repeatNum[0],'.');
+      removedIndexes[stripped] = newKeys[key];
+    }
+  });
+  return Object.assign(newKeys, removedIndexes);
+};
+
+const answerQuestions = async (answers, pathKeys,reportName) => {
+  answers.forEach((answer) => {
+    const answerKey = answer.name.replace(`/${reportName}/`,'').replace(/\//g,'.');
+    const answerVal = pathKeys[answerKey];
+    if (answeredQuestions.includes(answer.name) || !answerVal){
+      //already answered this question or there is no corresponding value in the report.
+      return;
+    }
+
+    const actionToExecute = answerActions[answer.type] || defaultAction;
+    actionToExecute(answer, answerVal);
+    answeredQuestions.push(answer.name);
+  });
+  // Need to see if there are any new answers since selecting a previous answer.
+  const possibleNewQuestions = await getVisibleAnswers();
+  if(possibleNewQuestions.some(x => !answeredQuestions.includes(x.name))) {
+    answerQuestions(possibleNewQuestions, pathKeys,reportName);
+  }
+};
+
+const fillForm = async (reportFields,reportName) => {
+  const flatObj = flatten(reportFields);
+  const pathKeys = repeats(flatObj);
+  const answers = await getVisibleAnswers();
+  answerQuestions(answers, pathKeys, reportName);
+
+  if(await helper.isDisplayed(genericFormPo.nextButton)) {
+    await genericFormPo.nextPageNative();
+    await fillForm(reportFields,reportName);
+  } else {
+    await genericFormPo.submitNative();
+  }
+};
+
+module.exports = {
+  fillForm
+};
```

Comment:
I'm a little uncomfortable with this as a pattern...

If I understand this correctly it tries to programmatically complete and submit the form so you don't have to explicitly list out all the fields yourself. This gets rid of a lot of boilerplate where we fill in different forms, but I worry that it could also obscure some bugs. One slightly obtuse example I can think of is if I wrote a bug that added 50 blank pages to the end of a form, I _think_ this code would copy with that, and the test would pass. The old way of doing it (eg: fill in 3 fields, hit next, fill in 2 fields, hit submit) would fail and catch the bug.

Reply:
I've changed this to take a array of form pages with defined css selectors and types. This constrains the form filling to be a static length. There is a bit more work to be done to finish it but wanted your thoughts before moving forward to deeply. Otherwise I can trash this form filling code and define large form filling functions in the specific form page object.